code/model/DMSDocumentSet.php

```diff
@@ -150,4 +170,98 @@ protected function addRequirements()
+                /** @var ListboxField $field */
+                $field->setMultiple(true)->setSource($membersMap);
+            }
+        }
+        $keyValPairs = new JsonField('KeyValuePairs', $dmsDocFields->toArray());
+
+        // Now lastly add the sort fields
+        $sortedBy = new FieldGroup("SortedBy", array(
+                new DropdownField("SortBy", '', array(
+                    'LastEdited'  => "Last changed",
+                    'LastChanged' => "Last replaced",
+                    'Created'     => "Created",
+                    'Title'       => "Document title",
+                ), 'LastEdited'),
+                new DropdownField("SortByDirection", '', $this->dbObject("SortByDirection")->enumValues(), "DESC"),
+            ));
+
+        $sortedBy->setTitle("Sort the document set by:");
+        $fields->addFieldsToTab('Root.QueryBuilder', $keyValPairs);
+        $fields->addFieldToTab('Root.QueryBuilder', $sortedBy);
+        $this->extend('updateQueryFields', $fields);
+    }
+
+    public function onBeforeWrite(){
+        parent::onBeforeWrite();
+
+        $this->saveLinkedDocuments();
+    }
+
+    /**
+     * Retrieve a list of the documents in this set. An extension hook is provided before the result is returned.
+     *
+     * @return ArrayList|null
+     */
+    public function saveLinkedDocuments()
+    {
+        // Documents that belong to just this set.
+        /** @var ManyManyList $originals */
+        $originals = $this->Documents();
+        if (!(empty($this->KeyValuePairs)) && $this->isChanged('KeyValuePairs')) {
+
+            $keyValuesPair = Convert::json2array($this->KeyValuePairs);
+            /** @var DMSDocument $dmsDoc */
+            $dmsDoc = singleton(DMSDocumentSet::$linked_document_type);
+            $context = $dmsDoc->getDefaultSearchContext();
+
+            $sortBy = $this->SortBy ? $this->SortBy : 'LastEdited';
+            $sortByDirection = $this->SortByDirection ? $this->SortByDirection : 'DESC';
+            $sortedBy = sprintf('%s %s', $sortBy, $sortByDirection);
+            /** @var DataList $documents */
+            $documents = $context->getResults($keyValuesPair, $sortedBy);
+            $now = SS_Datetime::now()->Rfc2822();
+            $documents = $documents->where(
+                "\"EmbargoedIndefinitely\" = 0 AND ".
+                " \"EmbargoedUntilPublished\" = 0 AND ".
+                "(\"EmbargoedUntilDate\" IS NULL OR " .
+                "(\"EmbargoedUntilDate\" IS NOT NULL AND '{$now}' >= \"EmbargoedUntilDate\")) AND " .
+                "\"ExpireAtDate\" IS NULL OR (\"ExpireAtDate\" IS NOT NULL AND '{$now}' < \"ExpireAtDate\")");
+
+            // Remove all BelongsToSet as the rules have changed
+            $originals->removeByFilter('"BelongsToSet" = 0');
+            foreach ($documents as $document) {
+                $originals->add($document, array('BelongsToSet' => 0));
```

Comment:
See ManyManyList::add for the second argument @robbie. It sets many_many_extraFields values.